odoo/tools/paper_muncher.py

```diff
@@ -0,0 +1,129 @@
+        subprocess.Popen([binary, '--version'], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        return True
+    except Exception as e:  # noqa: BLE001
+        _logger.warning(e)
+        return False
+
+
+def run_papermuncher(
+    paperformat,
+    bodies: list[str],
+    header: str = '',
+    footer: str = '',
+    landscape: bool = False,
+    specific_paperformat_args: dict | None = None,
+    set_viewport_size: str | None = None,
+) -> bytes:
+
+    # TODO: use header and footer (+ placeholder for pager)
+    def get_body(html):
+        return html and html.split('<body')[1].split(">", 1)[1].split("</body>")[0] or ''
+    header = get_body(header)
+    footer = get_body(footer)
+    content = ''
+    for html in bodies:
+        open, body, close = html.partition(get_body(html))
+        content += f'{open}{header}{body}{footer}{close}\n'
+
+
+    extra_args = ['--scale', '72dpi']  # wkhtml default is 72dpi
+    if landscape:
+        extra_args += ['--orientation', 'landscape'] 
+    if paperformat and paperformat.format:
+        if paperformat.format != 'custom':
+            extra_args += ['--paper', str(paperformat.format)]
+        elif paperformat.page_height and paperformat.page_width:  
+            extra_args += ['--width', str(paperformat.page_width) + 'mm']
+            extra_args += ['--height', str(paperformat.page_height) + 'mm']
+
+    with subprocess.Popen(
+        [find_in_path('paper-muncher'), '--sandbox', "print", REPORT_ASSET_KEY, '-o', 'http://stdout'] + extra_args,
+        stdin=subprocess.PIPE,
+        stdout=subprocess.PIPE,
+    ) as process:
+        # phase 1: send the html
+        read_pm_request(process.stdout) # throw the first request away
+        response_headers = (
+            f"HTTP/1.1 200 OK\r\n"
+            f"Content-Length: {len(content.encode())}\r\n"
+            f"Content-Type: text/html\r\n\r\n"
+        )
+        process.stdin.write(response_headers.encode() + content.encode())
+
+        # phase 2: send the assets
+        while (path := read_pm_request(process.stdout)) is not None:
+            for chunk in mock_request(path):
+                process.stdin.write(chunk)
+                process.stdin.flush()
+
+        # phase 3: fetch the pdf
+        process.stdin.write(b"HTTP/1.1 200 OK\r\n\r\n")
+        process.stdin.flush()
+        rendered_content = process.stdout.read()
+        process.terminate()
+        return rendered_content
+
+
+def read_pm_request(stdout):
+    while True:
+        request_line = stdout.readline().decode('utf-8')
+        if len(request_line) == 0:
+            raise EOFError("Input stream has ended")
+        if request_line == "\r\n":
+            break
+        if request_line.startswith('PUT'):
+            path = None
+        elif request_line.startswith('GET'):
+            parsed_line = request_line.split(' ', 2)
+            if len(parsed_line) != 3:
+                raise ValueError("Invalid request line")
```

Comment:
achieve exactly the same, but in a single line :nerd_face: 
```suggestion
            _method, path = request_line.split(' ')
```